text/3697-declarative-attribute-macros.md

```diff
@@ -0,0 +1,123 @@
+- Feature Name: `declarative_attribute_macros`
+- Start Date: 2024-09-20
+- RFC PR: [rust-lang/rfcs#3697](https://github.com/rust-lang/rfcs/pull/3697)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Support defining `macro_rules!` macros that work as attribute macros.
+
+# Motivation
+[motivation]: #motivation
+
+Many crates provide attribute macros. Today, this requires defining proc
+macros, in a separate crate, typically with several additional dependencies
+adding substantial compilation time, and typically guarded by a feature that
+users need to remember to enable.
+
+However, many common cases of attribute macros don't require any more power
+than an ordinary `macro_rules!` macro. Supporting these common cases would
+allow many crates to avoid defining proc macros, reduce dependencies and
+compilation time, and provide these macros unconditionally without requiring
+the user to enable a feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a `macro_rules!` macro, you can prefix some of the macro's rules
+with `attribute(...) =>` to allow using the macro as an attribute. The
+arguments to the attribute, if any, are parsed by the *MacroMatcher* in the
+first set of parentheses; the second *MacroMatcher* parses the entire construct
+the attribute was applied to. The resulting macro will work anywhere an
+attribute currently works.
+
+```rust
+macro_rules! main {
+    attribute() => ($func:item) => { make_async_main!($func) };
+    attribute(threads = $threads:literal) => ($func:item) => { make_async_main!($threads, $func) };
+}
+
+#[main]
+async fn main() { ... }
+
+#[main(threads = 42)]
+async fn main() { ... }
+```
+
+Attribute macros defined using `macro_rules!` follow the same scoping rules as
+any other macro, and may be invoked by any path that resolves to them.
+
+An attribute macro must not require itself for resolution, either directly or
+indirectly (e.g. applied to a containing module or item).
+
+Note that a single macro can have both attribute and non-attribute rules.
```

Comment:
This is something that a proc-macro can't do, isn't it? Therefore, it seems to me that it would be wise to not allow it, to keep API and implementation less potentially coupled.

I know it is a common practice for `macro_rules` macros to have additional rules where they call themselves, but they can always use a second macro (referenced via `$crate`) if needed.

Reply:
@kpreid It'd be trivial to disallow it, and I can add that as an unresolved question. However, in addition to wanting to avoid separate helper macros (and reuse common existing patterns of `macro_rules!` macros), I have an additional reason for proposing this:

In some cases, existing crates may already have a macro with a given name, and want to use the *same* name (rather than a different one) for their attribute macro. For instance, consider a crate like `smol` that has provided a `main` `macro_rules!` macro for use with https://crates.io/crates/macro_rules_attribute, and now wants to expose it as an attribute macro as well; they shouldn't have to call that attribute something *other* than `main`.

Is there any particular reason to expect that this functionality would cause a problem?

Reply:
> Is there any particular reason to expect that this functionality would cause a problem?

Suppose a library has such a macro, and the maintainers want to add a feature, but the feature cannot be expressed within `macro_rules!` and they wish to replace the macro with a proc-macro, without making any semver-breaking changes. They can't.

Oh, but this is solvable by having the `macro_rules` macro *expand to a proc-macro attribute*, isn't it.

Reply:
Right, exactly.